crates/subspace-networking/src/utils/piece_provider.rs

```diff
@@ -136,6 +136,24 @@ where
         retry(backoff, || async {
             let current_attempt = retries.fetch_add(1, Ordering::Relaxed);
 
+            // Wait until we connect to DSN.
+            let mut dsn_connection_observer = self.node.dsn_connection_observer().clone();
+            loop {
+                let dsn_connected = *dsn_connection_observer.borrow();
+
+                if !dsn_connected {
+                    debug!(%piece_index, current_attempt, "Couldn't get a piece from DSN. No DSN connection...");
+
+                    if let Err(err) = dsn_connection_observer.changed().await {
```

Comment:
This is racy, state can change between reading and subscription. This whole loop needs to be rewritten. Subscription should be created first, then state checked, then you wait for changes and check in a loop for current value.

Reply:
Multiple `false` values is unfortunate, I would have expected it to ignore updates if new value is the same as latest value. It doesn't have any bound on `T` (like `PartialEq`) so I guess you're right, it doesn't care. Though with single source of updates you kind of guarantee it anyway.

In general semantics of receivers is that they give you updates since the point in time when you create the receiver. So if you check current status first and later subscribe to updates, state may change in between those two. This is a classical bug, I believe @liuchengxu or @NingLin-P just discovered one of such bugs earlier this week with one of subscriptions.

So what could happen is this:
* check is made, node is offline
* node goes online
* subscription for changes is created
* waiting for changes... for a very long time
* node goes offline
* above subscription gets a notification about some change
* another iteration of the loop
* node is still offline

As the result a lot of time wasted unnecessarily.

Reply:
50 lines with 4 channels interleaved is hard to read, what do you try to explain by that or what is not clear in my above comment? What you have is a classic race condition that needs to be fixed. Even if it happens to accidentally work, it still needs to be fixed so it is obvious to the reader that there is no issue here.

Reply:
I cleaned the noise in the code snippet (it's half in size now). Please, have another look: https://gist.github.com/shamil-gadelshin/8ec1ad8f2f971cab7afba248c1aa868f
I don't see the step of "long waiting" neither in the `watch::channel` code nor in the `emulation` from the gist.